### PR TITLE
Fixed a bug that may cause a job to hang if its tasks fail to getExtractor

### PR DESCRIPTION
**Pull request by liyinan926**
This bug is affecting stand-alone jobs using `LocalJobLauncher` that may hang if any of its tasks fails to `getExtractor`. In this case, the `finally` part of the `Task.run` method will be executed, and `this.taskStateTracker.onTaskCompletion(this)` is called. In the implementation class `LocalTaskStateTracker2`, `this.taskExecutor.retry(task)` is called, which will throw an exception on `task.getTaskState().getPropAsInt(ConfigurationKeys.FORK_BRANCHES_KEY)` since the key `FORK_BRANCHES_KEY` is not set because the task failed before the number of branches is known and that key is set. This runtime exception is thrown in the `finally` clause and is not caught, causing the task to hang.

Signed-off-by: Yinan Li <liyinan926@gmail.com>

### CODE REVIEW

**Comment by liyinan926** [comment]
@zliu41 can you review this?

File: gobblin-runtime/src/test/resources/log4j.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (zliu41) [inline gobblin-runtime/src/test/resources/log4j.xml]: Why is this needed? There's a `log4j.xml` in `conf/`, is it possible to reuse that one?

**Comment by zliu41** [comment]
In `LocalTaskStateTracker2.onTaskCompletion`, should `task.markTaskCompletion();` be moved to a finally block to make sure it is called?

File: gobblin-runtime/src/test/resources/log4j.xml (hunk 1 above)
Comment (liyinan926) [inline gobblin-runtime/src/test/resources/log4j.xml]: It is slightly different from that one as this is only used for testing.

**Comment by liyinan926** [comment]
`task.markTaskCompletion` should only be called if all retries fail. So calling it in a finally clause is not appropriate. What we can do is to make sure it is called if any `Throwable` is thrown.

**Comment by zliu41** [comment]
So should it be put in a `catch(Throwable t)` block?

**Comment by liyinan926** [comment]
Yes, updated.

**Comment by zliu41** [comment]
LGTM
